subprojects/diagnostics/src/main/java/org/gradle/api/tasks/diagnostics/internal/graph/nodes/SelectionReasonHelper.java

```diff
@@ -20,7 +20,20 @@
 
 public abstract class SelectionReasonHelper {
     public static String getReasonDescription(ComponentSelectionReason reason) {
-        if (reason.isExpected() && !((ComponentSelectionReasonInternal) reason).hasCustomDescriptions()) {
+        ComponentSelectionReasonInternal r = (ComponentSelectionReasonInternal) reason;
+        String description = getReasonDescription(r);
+        if (reason.isConstrained()) {
+            if (description == null || !r.hasCustomDescriptions()) {
```

Comment:
It doesn't make sense to me why we have to check `!r.hasCustomDescriptions()` here. I would expect `getReasonDescription()` to return null if there's no reason worth showing.
I suspect the issue is that `reason.isExpected() == false` for a dependency constraint, and I'm not sure that's correct.